gems/smithy-schema/lib/smithy-schema/document.rb

```diff
@@ -0,0 +1,104 @@
+# frozen_string_literal: true
+
+require_relative 'document_utils'
+
+module Smithy
+  module Schema
+    # A Smithy document type, representing typed or untyped data from Smithy data model.
+    # ## Document types
+    # Document types are protocol-agnostic view of untyped data. They could be combined
+    # with a schema to serialize its contents.
+    #
+    # Smithy-Ruby currently only support JSON documents.
+    class Document
+      # @param  [Object] data  document data
+      # @param [Hash] options
+      # @option options [Smithy::Schema::Structure] :schema schema to reference when setting
+      #  document data. Only applicable when data param is a type of {Shapes::StructureShape}.
+      # @option options [Boolean] :use_timestamp_format Whether to use the `timestampFormat`
+      #  trait or ignore it when creating a {Document} with given schema. The `timestampFormat`
+      #  trait is ignored by default.
+      # @option options [Boolean] :use_json_name Whether to use the `jsonName` trait or ignore
+      #  it when creating a {Document} with given schema. The `jsonName` trait is ignored
+      #  by default.
+      def initialize(data, options = {})
+        @data = set_data(data, options)
+        @discriminator = extract_discriminator(data, options)
+      end
+
+      # @return [Object] data
+      attr_reader :data
+
+      # @return [String] discriminator
+      attr_reader :discriminator
+
+      # @param [Object] key
+      # @return [Object]
+      def [](key)
+        return unless @data.is_a?(Hash) && @data.key?(key)
+
+        @data[key]
+      end
+
+      # @param [Shapes::Shape] schema
+      # @return [Shapes::Structure] typed shape
+      def as_typed(schema)
+        error_message = 'Invalid schema or document data'
+        raise ArgumentError, error_message unless valid_schema?(schema) && @data.is_a?(Hash)
+
+        type = schema.type.new
+        DocumentUtils.apply(@data, schema, type)
+      end
+
+      private
+
+      def discriminator?(data)
+        data.is_a?(Hash) && data.key?('__type')
+      end
+
+      def extract_discriminator(data, opts)
+        return if data.nil?
+
+        return unless discriminator?(data) || (schema = opts[:schema])
+
+        if discriminator?(data)
+          data['__type']
+        else
+          error_message = "Expected a structure schema, given #{schema.class} instead"
+          raise error_message unless valid_schema?(schema)
+
+          schema.id
+        end
+      end
+
+      def set_data(data, opts)
+        return if data.nil?
+
+        case data
+        when Smithy::Schema::Structure
```

Comment:
Isn't this already "done"? If it's a structure then it's already typed.

Reply:
Can you expand on that? what do you mean by being "done" - I still want to re-format the typed shape into a document format.